photon-api/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala

```diff
@@ -194,25 +176,46 @@ class RandomEffectModel(
       val areTypesEqual = this.randomEffectType == other.randomEffectType
       val areShardsEqual = this.featureShardId == other.featureShardId
       lazy val areAllModelsEqual = this
-        .modelsRDD
-        .fullOuterJoin(other.modelsRDD)
-        .mapPartitions { iterator =>
-
-          val areModelsEqual = iterator.forall {
-            case (_, (Some(model1), Some(model2))) => model1.equals(model2)
-            case _ => false
-          }
-
-          Iterator.single(areModelsEqual)
-        }
-        .fold(true)(_ && _)
+        .models
+        .withColumnRenamed(DataConst.COEFFICIENTS, "s1")
+        .join(other.models.withColumnRenamed(DataConst.COEFFICIENTS, "s2"), col(DataConst.ID), "fullouter")
+        .filter("s1 is null or s2 is null or s1 != s2") //TODO: add udf to compare two vectors
+        .head(1)
+        .isEmpty
 
       areTypesEqual && areShardsEqual && areAllModelsEqual
 
     case _ =>
       false
   }
 
+  /**
+   * Convert models from dataframe to RDD
+   * @return
+   */
+  def toRDD(): RDD[(REType, GeneralizedLinearModel)] = {
```

Comment:
This function shouldn't exist - the join should occur between `DataFrame` objects, and then the resulting `DataFrame` should be converted to an `RDD`.